### PR TITLE
haskell-ledger-bindings: TransactionService, GetTransactions

### PR DESCRIPTION
**Pull request by nickchapman-da**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/hs/bindings/BUILD.bazel [hunk 1]
```diff
@@ -9,7 +9,10 @@ da_haskell_library(
     hazel_deps = [
         "base",
         "text",
+        "vector",
         "containers",
+        "proto3-suite",
+        #"proto3-wire",
```
Comment (neil-da) [inline language-support/hs/bindings/BUILD.bazel]: No point checking in commented out dependencies

File: language-support/hs/bindings/examples/View.hs [hunk 2]
```diff
@@ -1,15 +1,40 @@
...
 import           System.Environment
 
 main :: IO ()
 main = do
-    putStrLn "Ledger App (View)..."
-    [port] <- getArgs
-    id <- LedgerIdentity.ledgerId (Port (read port))
-    print (port, id)
+    port <- fmap (Port . read . head) $ getArgs
+    putStrLn$ "Connecting to ledger on port: " ++ show port
```
Comment (neil-da) [inline language-support/hs/bindings/examples/View.hs]:
```suggestion
    putStrLn $ "Connecting to ledger on port: " ++ show port
```

Everyone always puts spaces before and after `$`. Your `putStrLn$` gives me flash backs to visual basic.

File: language-support/hs/bindings/examples/View.hs [hunk 3]
```diff
@@ -1,15 +1,40 @@
...
+    putStrLn "Getting Transaction streams."
+    let alice = Party "Alice"
+    let bob = Party "Bob"
+    aliceTs <- Ledger.getTransactionStream h alice
+    bobTs <- Ledger.getTransactionStream h bob
+    putStrLn "Watching Transaction streams."
+    watch alice aliceTs
+    watch bob bobTs
+    waitForALongTime
+    return ()
```
Comment (neil-da) [inline language-support/hs/bindings/examples/View.hs]: This return is redundant

File: language-support/hs/bindings/examples/View.hs [hunk 4]
```diff
@@ -1,15 +1,40 @@
...
+
+watch :: Show a => Party -> ResponseStream a -> IO ()
+watch party rs = do _ <- forkIO $ loop (1::Int); return ()
```
Comment (neil-da) [inline language-support/hs/bindings/examples/View.hs]:
```suggestion
watch party rs = void $forkIO $ loop (1::Int)
```

File: language-support/hs/bindings/examples/View.hs [hunk 5]
```diff
@@ -1,15 +1,40 @@
...
+    where loop n = do
+              x <- nextResponse rs
+              putStrLn $ show party <> "(" <> show n <> ") = " <> show x
+              loop (n+1)
 
+waitForALongTime :: IO ()
+waitForALongTime = do
+    putStrLn "Pause for a year"
+    threadDelay (10 ^ (6 :: Int) * 3600 * 24 * 365)
```
Comment (neil-da) [inline language-support/hs/bindings/examples/View.hs]: Use System.Time.Extra.Sleep, and then you can work in seconds. But if you just want to wait "forever" a `getLine` might be reasonable - terminate when the user hits enter.

File: language-support/hs/bindings/src/LedgerHL.hs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+data LedgerHandle = LedgerHandle { port :: Port, id :: LedgerId }
+
+identity :: LedgerHandle -> LedgerId
+identity LedgerHandle{id} = id
+
+
+newtype Party = Party { party :: Text }
+
+instance Show Party where
+    show Party{party} = Text.unpack party
```
Comment (neil-da) [inline language-support/hs/bindings/src/LedgerHL.hs]:
```suggestion
    show = Text.unpack . party
```

File: language-support/hs/bindings/src/LedgerHL.hs [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    log$ "ledgerIdentity"
+    withGRPCClient (config port) $ \client -> do
+        rpcs <- LIS.ledgerIdentityServiceClient client
+        callLedgerIdService rpcs
+
+callLedgerIdService :: LIS.LedgerIdentityService ClientRequest ClientResult -> IO LedgerId
+callLedgerIdService (LIS.LedgerIdentityService rpc) = do
+    response <- rpc (wrap (LIS.GetLedgerIdentityRequest noTrace))
+    LIS.GetLedgerIdentityResponse text <- unwrap response
+    return$ LedgerId text
```
Comment (neil-da) [inline language-support/hs/bindings/src/LedgerHL.hs]:
```suggestion
    return $ LedgerId text
```

File: language-support/hs/bindings/examples/View.hs (hunk 2 above)
Comment (shayne-fletcher) [reply language-support/hs/bindings/examples/View.hs]: More generally, we hashed it out and advocate white space around all binary operators, notably including cons e.g `x : xs`.

File: language-support/hs/bindings/examples/View.hs (hunk 4 above)
Comment (garyverhaegen-da) [reply language-support/hs/bindings/examples/View.hs]: Should there be a space there too? `void $ forkIO`?

File: language-support/hs/bindings/src/LedgerHL.hs [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+
+connect :: Port -> IO LedgerHandle
+connect port = do
+    id <- getLedgerIdentity port
+    return$ LedgerHandle {port, id}
+
+getLedgerIdentity :: Port -> IO LedgerId
+getLedgerIdentity port = do
+    log$ "ledgerIdentity"
```
Comment (garyverhaegen-da) [inline language-support/hs/bindings/src/LedgerHL.hs]: This line fails the build; should be
```suggestion
    log "ledgerIdentity"
```
according to the linter

File: language-support/hs/bindings/src/LedgerHL.hs [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+newtype ResponseStream a = ResponseStream { chan :: Chan a }
+
+nextResponse :: ResponseStream a -> IO a
+nextResponse ResponseStream{chan} = readChan chan
+
+
+connect :: Port -> IO LedgerHandle
+connect port = do
+    id <- getLedgerIdentity port
+    return$ LedgerHandle {port, id}
```
Comment (garyverhaegen-da) [inline language-support/hs/bindings/src/LedgerHL.hs]: Assuming this needs a space too, for consistency with Neil's other remarks.

File: language-support/hs/bindings/examples/View.hs [hunk 10]
```diff
@@ -1,15 +1,40 @@
...
 main = do
-    putStrLn "Ledger App (View)..."
-    [port] <- getArgs
-    id <- LedgerIdentity.ledgerId (Port (read port))
-    print (port, id)
+    port <- fmap (Port . read . head) $ getArgs
+    putStrLn$ "Connecting to ledger on port: " ++ show port
+    h <- Ledger.connect port
+    let id = Ledger.identity h
+    putStrLn$ "LedgerIdentity = " <> show id
```
Comment (garyverhaegen-da) [inline language-support/hs/bindings/examples/View.hs]: Space?

File: language-support/hs/bindings/examples/View.hs [hunk 11]
```diff
@@ -1,15 +1,40 @@
...
+watch :: Show a => Party -> ResponseStream a -> IO ()
+watch party rs = do _ <- forkIO $ loop (1::Int); return ()
+    where loop n = do
+              x <- nextResponse rs
+              putStrLn $ show party <> "(" <> show n <> ") = " <> show x
+              loop (n+1)
 
+waitForALongTime :: IO ()
+waitForALongTime = do
+    putStrLn "Pause for a year"
```
Comment (garyverhaegen-da) [inline language-support/hs/bindings/examples/View.hs]: If using the `getLine` as suggested by Neil, may be worth changing this to
```suggestion
    putStrLn "Press Enter to continue"
```

File: language-support/hs/bindings/examples/View.hs (hunk 2 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/examples/View.hs]: ok

File: language-support/hs/bindings/examples/View.hs (hunk 4 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/examples/View.hs]: Gary, you are a tease :)

File: language-support/hs/bindings/src/LedgerHL.hs (hunk 8 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/src/LedgerHL.hs]: yup

File: language-support/hs/bindings/examples/View.hs (hunk 11 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/examples/View.hs]: thanks

**Pull request merged**